python/lsst/pipe/base/executionButlerBuilder.py

```diff
@@ -39,6 +40,52 @@
 DataSetTypeMap = Mapping[DatasetType, Set[DataCoordinate]]
 
 
+def _validate_dataset_type(
+    candidate: DatasetType, previous: dict[Union[str, DatasetType], DatasetType]
+) -> DatasetType:
+    """Check the dataset types and return a consistent variant if there are
+    different compatible options.
+
+    Parameters
+    ----------
+    candidate : `lsst.daf.butler.DatasetType`
+        The candidate dataset type.
+    previous : `dict` [Union[`str`, `DatasetType`], `DatasetType`]
+        Previous dataset types found, indexed by name and also by
+        dataset type. The latter provides a quick way of returning a
+        previously checked dataset type.
+
+    Returns
+    -------
+    datasetType : `lsst.daf.butler.DatasetType`
+        The dataset type to be used.
+    """
+    # First check that if we have previously vetted this dataset type.
+    checked = previous.get(candidate)
+    if checked:
+        return checked
+
+    # Have not previously encountered this dataset type.
+    name = candidate.name
+    if prevDsType := previous.get(name):
+        # Check compatibility. For now assume both directions have to
+        # be acceptable.
+        if prevDsType.is_compatible_with(candidate) and candidate.is_compatible_with(prevDsType):
+            # Store this mapping and declare that we will use it in the future.
+            previous[candidate] = prevDsType
+            # Use the first type encountered.
+            return prevDsType
+
+        raise ValueError(
+            f"Dataset type incompatibility in graph: {prevDsType} not compatible with {candidate}"
+        )
+
+    # New dataset type.
+    previous[name] = candidate
+    previous[candidate] = candidate
```

Comment:
Is `previous` persisted somehow? I don't immediately see the benefit to updating it here.

Reply:
It's used in the initial look up next time it gets called. There is a short circuit at the start of this function that sees if the dataset type has previously been handled and if it does exist it returns the value immediately. This saves having to do the is_compatible_with and == checks thousands of times as the graph is processed. I'm being a bit cute in storing keys of both DatasetType and the dataset type name in the same dict.  This is all done in line 64 above. It maps the candidate dataset type to a previously checked one and returns it if it already knows what to do. I will try to expand the function doc string.